lib/util/http/bitbucket-server.ts

```diff
@@ -1,30 +1,79 @@
-import { resolveBaseUrl } from '../url';
+import is from '@sindresorhus/is';
+import { logger } from '../../logger';
+import { parseUrl, resolveBaseUrl } from '../url';
 import type { HttpOptions, HttpResponse, InternalHttpOptions } from './types';
 import { Http } from '.';
 
+const MAX_LIMIT = 100;
+
 let baseUrl: string;
 export const setBaseUrl = (url: string): void => {
   baseUrl = url;
 };
 
-export class BitbucketServerHttp extends Http {
+export interface BitbucketServerHttpOptions extends HttpOptions {
+  paginate?: boolean;
+  limit?: number;
+}
+
+interface PagedResult<T = unknown> {
+  nextPageStart?: number;
+  values: T[];
+}
+
+export class BitbucketServerHttp extends Http<BitbucketServerHttpOptions> {
   constructor(options?: HttpOptions) {
     super('bitbucket-server', options);
   }
 
-  protected override request<T>(
+  protected override async request<T>(
     path: string,
-    options?: InternalHttpOptions,
+    options?: InternalHttpOptions & BitbucketServerHttpOptions,
   ): Promise<HttpResponse<T>> {
-    const url = resolveBaseUrl(baseUrl, path);
-    const opts = {
-      baseUrl,
-      ...options,
-    };
+    const opts = { baseUrl, ...options };
     opts.headers = {
       ...opts.headers,
       'X-Atlassian-Token': 'no-check',
     };
-    return super.request<T>(url, opts);
+
+    const resolvedUrl = parseUrl(resolveBaseUrl(baseUrl, path));
+    if (!resolvedUrl) {
+      logger.error({ path }, 'Bitbucket Server: cannot parse path');
+      throw new Error(`Bitbucket Server: cannot parse path ${path}`);
+    }
+
+    if (opts.paginate) {
+      const limit = opts.limit ?? MAX_LIMIT;
+      resolvedUrl.searchParams.set('limit', limit.toString());
+    }
+
+    const result = await super.request<T | PagedResult<T>>(
+      resolvedUrl.toString(),
+      opts,
+    );
+
+    if (opts.paginate && isPagedResult(result.body)) {
+      const collectedValues: T[] = [...result.body.values];
```

Comment:
```suggestion
      const collectedValues = [...result.body.values];
```
I think we don't need this extra type